ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt

```diff
@@ -119,13 +121,71 @@ internal class Endpoint(
             setupTimeout(callContext, request, timeout)
 
             val requestTime = GMTDate()
-            writeRequest(request, output, callContext, proxy != null)
-            return readResponse(requestTime, request, input, originOutput, callContext)
+            val overProxy = proxy != null
+
+            return if (expectContinue(request.headers[HttpHeaders.Expect], request.body)) {
+                processExpectContinue(request, input, output, originOutput, callContext, requestTime, overProxy)
+            } else {
+                writeRequest(request, output, callContext, overProxy)
+                readResponse(requestTime, request, input, originOutput, callContext)
+            }
         } catch (cause: Throwable) {
             throw cause.mapToKtor(request)
         }
     }
 
+    @OptIn(InternalCoroutinesApi::class)
+    private suspend fun processExpectContinue(
+        request: HttpRequestData,
+        input: ByteReadChannel,
+        output: ByteWriteChannel,
+        originOutput: ByteWriteChannel,
+        callContext: CoroutineContext,
+        requestTime: GMTDate,
+        overProxy: Boolean,
+    ) = withContext(callContext) {
+        writeHeaders(request, output, overProxy)
+
+        val response = withTimeoutOrNull(CONTINUE_RESPONSE_TIMEOUT_MILLIS) {
+            val job = Job()
+            coroutineContext.job.invokeOnCompletion(onCancelling = true) {
+                if (it == null) job.complete()
+                else job.completeExceptionally(it)
+            }
+            callContext.job.invokeOnCompletion {
+                if (it == null) job.complete()
+                else job.completeExceptionally(it)
+            }
+
+            readResponse(requestTime, request, input, originOutput, callContext + job)
```

Comment:
It's better to have `awaitContent` instead of `readResponse` inside `withTimeoutOrNull`. otherwise, you can consume only part of the response until the timeout is reached and the channel will have partial data remaining.